samples/cpp/gstreamer_pipeline.cpp

```diff
@@ -0,0 +1,205 @@
+#include "opencv2/core/utility.hpp"
+#include "opencv2/imgproc.hpp"
+#include "opencv2/imgcodecs.hpp"
+#include "opencv2/highgui.hpp"
+#include <iostream>
+
+using namespace std;
+using namespace cv;
+
+string getGstDemuxPlugin(CommandLineParser &parser);
+string getGstAvDecodePlugin    (CommandLineParser &parser);
+void checkSupportArguments(CommandLineParser &parser);
+
+int main(int argc, char *argv[])
+{
+    const string keys =
+          "{h help usage ? |           | print help message   }"
+          "{s show         |false      | show frames }"
+          "{p pipeline     |gst-default| pipeline name  (supported: 'gst-default', 'gst-vaapi', 'gst-libav', 'ffmpeg') }"
+          "{ct container   |mp4        | container name (supported: 'mp4', 'mov', 'avi', 'mkv') }"
+          "{cd codec       |h264       | codec name     (supported: 'h264', 'h265', 'mpeg4', 'mjpeg', 'vp8') }"
+          "{f file path    |           | path to file }";
+
+    CommandLineParser parser(argc, argv, keys);
+
+    parser.about("This program shows how to read a video file with GStreamer pipeline with OpenCV.");
+
+    if (parser.has("help"))
+    {
+        parser.printMessage();
+        return 0;
+    }
+
+    if (!parser.check())
+    {
+        parser.printErrors();
+        return 0;
+    }
+
+    bool   arg_show;      // show frames on the screen
+    string arg_pipeline,  // GStreamer pipeline type
+           arg_container, // container type
+           arg_codec,     // codec type
+           arg_file_name, // path to videofile
+           codec_name,
+           meta_info;
+    VideoCapture cap;
+
+    arg_show      = parser.get<bool>  ("show");
+    arg_pipeline  = parser.get<string>("pipeline");
+    arg_container = parser.get<string>("container");
+    arg_codec     = parser.get<string>("codec");
+    arg_file_name = parser.get<string>("file");
+
+    // Check support plugins, containers and codecs
+    checkSupportArguments(parser);
+
+    // Choose the constructed GStreamer pipeline
+    if (arg_pipeline.find("gst") != string::npos)
+    {
+        ostringstream pipeline;
+        pipeline << "filesrc location=\"" << arg_file_name << "\"";
+        pipeline << " ! " << getGstDemuxPlugin(parser);
+
+        if (arg_pipeline.find("default") != string::npos)
+        {
+            pipeline << " ! decodebin";
```

Comment:
Maybe `gst-default` should mean opening video stream by filename:
```.cpp
cap = VideoCapture(arg_file_name, CAP_GSTREAMER);
```
Will it work?